### PR TITLE
Improved asset code (with realistic interfaces)

### PR DESCRIPTION
**Pull request by AurelienGasser**
As discussed with @jmorel here is an alternative implementation for the opener/algo/metrics.

They are meant to 1. have an implementation that is very simple and 2. use interfaces that match the intended usage, in particular:

- `train` returns a model structure
- `predict` calculates `y` based on the input `model` and the input `X` 

Please let me know your thoughts. I'm particularly interested in feedback on the comments,

## Performance impact

I compared the time it takes to run `make test` between:

- `master`
- `improved_asset_code`
- `improved_asset_code_v2`

I ran `make test` 10x on each branch. Both `improved_asset_code` and `improved_asset_code_v2` take longer to run, but the increase is very small.

image

(those numbers are not worth much, it's just to get a general idea of the size of the increase)

In a future iteration, we could make changes so that the new algo/opener/metrics are only used for a few specific tests that ensure that data flows correctly between opener, algo and metrics. All the other tests could use a dummy algo that does nothing (like in `master`). This would shave a few seconds :man_shrugging:

### CODE REVIEW

**Comment by samlesu** [comment]
@AurelienGasser could you review @jmorel's comment to move on this topic ?

**Comment by AurelienGasser** [comment]
@jmorel I finally had a chance to look at this.

I made changes, what do you think?

File: substratest/factory.py [hunk 1]
```diff
@@ -16,21 +16,32 @@
...
+        for folder in folders:
+            with open(os.path.join(folder, '{DEFAULT_DATA_SAMPLE_FILENAME}'), 'r') as f:
+                reader = csv.reader(f)
+                for row in reader:
+                    res.append(int(row[1]))
+        return res  # returns a list of 2's
+    def fake_X(self):
+        return [1]
+    def fake_y(self):
+        return [2]
```
Comment (jmorel) [inline substratest/factory.py:44]: There is a new n_samples arg for the fake_X and fake_y method, but this means the code could become 
```python
    def fake_X(self, n_samples=1):
        return [1] * n_samples
    def fake_y(self, n_samples=1):
        return [2] * n_samples
```

**Comment by jmorel** [comment]
I'm adding @Esadruhn as a reviewer because she also made an attempt at proper asset code.

File: substratest/factory.py (hunk 1 above)
Comment (AurelienGasser) [reply substratest/factory.py]: nice! done

**Comment by samlesu** [comment]
💯

**Pull request merged**